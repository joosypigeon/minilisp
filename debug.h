#ifndef DEBUG_H
#define DEBUG_H

#include <stdio.h>

// Define debug levels using an enum for clarity
// Define preprocessor macros for use in #if
#define DEBUG_LEVEL_NONE    0
#define DEBUG_LEVEL_ERROR   1
#define DEBUG_LEVEL_WARN    2
#define DEBUG_LEVEL_INFO    3
#define DEBUG_LEVEL_VERBOSE 4

// Default level if none is set before inclusion
#ifndef DEBUG_LEVEL
#define DEBUG_LEVEL DEBUG_LEVEL_NONE
#endif

#if DEBUG_LEVEL >= DEBUG_LEVEL_ERROR
    #define DEBUG_PRINT_ERROR(...)   printf("ERROR: " __VA_ARGS__)
#else
    #define DEBUG_PRINT_ERROR(...)   ((void)0)
#endif

#if DEBUG_LEVEL >= DEBUG_LEVEL_WARN
    #define DEBUG_PRINT_WARN(...)    printf("WARN: " __VA_ARGS__)
#else
    #define DEBUG_PRINT_WARN(...)    ((void)0)
#endif

#if DEBUG_LEVEL >= DEBUG_LEVEL_INFO
    #define DEBUG_PRINT_INFO(...)    printf("INFO: " __VA_ARGS__)
#else
    #define DEBUG_PRINT_INFO(...)    ((void)0)
#endif

#if DEBUG_LEVEL >= DEBUG_LEVEL_VERBOSE
    #define DEBUG_PRINT_VERBOSE(...) printf("VERBOSE: " __VA_ARGS__)
#else
    #define DEBUG_PRINT_VERBOSE(...) ((void)0)
#endif

#define _STR_HELPER(x) #x
#define _STR(x) _STR_HELPER(x)

#endif // DEBUG_H
