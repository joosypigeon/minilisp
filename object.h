#ifndef OBJECT_H
#define OBJECT_H

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include "util.h"

#define SYM_INT     "int"
#define SYM_SYMBOL  "symbol"
#define SYM_LAMBDA  "lambda"
#define SYM_PAIR    "pair"
#define SYM_NIL     "nil"
#define SYM_UNKOWN  "unkown"

#define SYM_DEFINE  "define"
#define SYM_QUOTE   "quote"
#define SYM_EQUAL   "equal?"
#define SYM_PLUS    "+"
#define SYM_TIMES   "*"
#define SYM_CHAR_QUOTE  '\''
#define SYM_SPACE  ' '

typedef enum {
    TYPE_INT,
    TYPE_SYMBOL,
    TYPE_LAMBDA,
    TYPE_PAIR,
    TYPE_NIL
} Type;

typedef struct Object Object;

typedef struct Env Env; // forward declaration for closure support

struct Object {
    Type type;
    union {
        int int_val;
        char *symbol;

        struct {
            Object *params;
            Object *body;
            Env *env;
        } lambda;

        struct {
            Object *car;
            Object *cdr;
        }; // for TYPE_PAIR
    };
};

// Singleton NIL object
static Object *NIL = &(Object){ .type = TYPE_NIL };

// Singleton true object
static Object *TRUE = &(Object){ .type = TYPE_SYMBOL, .symbol = "#t" };

bool islist(Object *item){
    return item->type==TYPE_PAIR;
}

Object *make_number(int n) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_INT;
    obj->int_val = n;
    return obj;
}

Object *make_symbol(const char *name) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_SYMBOL;
    obj->symbol = xstrdup(name);
    return obj;
}

Object *make_lambda(Object *params, Object *body, Env *env) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_LAMBDA;
    obj->lambda.params = params;
    obj->lambda.body = body;
    obj->lambda.env = env;
    return obj;
}

Object *make_nil(void) {
    return NIL;
}

Object *cons(Object *car, Object *cdr) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_PAIR;
    obj->car = car;
    obj->cdr = cdr;
    return obj;
}

Object *car(Object *obj) {
    if (obj->type != TYPE_PAIR) {
        fprintf(stderr, "Error: car called on non-pair\n");
        exit(1);
    }
    return obj->car;
}

Object *cdr(Object *obj) {
    if (obj->type != TYPE_PAIR) {
        fprintf(stderr, "Error: cdr called on non-pair\n");
        exit(1);
    }
    return obj->cdr;
}

Object *cadr(Object *obj) {
    return car(cdr(obj));
}

Object *caddr(Object *obj) { return car(cdr(cdr(obj))); }

char * type_to_string(Type type) {
    char *s = NULL;
    switch (type) {
        case TYPE_INT:    s = xstrdup(SYM_INT);
        case TYPE_SYMBOL: s = xstrdup(SYM_SYMBOL);
        case TYPE_LAMBDA: s = xstrdup(SYM_LAMBDA);
        case TYPE_PAIR:   s = xstrdup(SYM_PAIR);
        case TYPE_NIL:    s = xstrdup(SYM_NIL);
        default:          s = xstrdup(SYM_UNKOWN);
    }
    return s;
}

void print_type(Type type) {
    printf("%s", type_to_string(type));
}

void print_object(Object *obj) {
    switch (obj->type) {
        case TYPE_INT:
            printf("%d", obj->int_val);
            break;
        case TYPE_SYMBOL:
            printf("%s", obj->symbol);
            break;
        case TYPE_LAMBDA:
            printf("(lambda ");
            print_object(obj->lambda.params);
            printf(" ");
            print_object(obj->lambda.body);
            printf(")");
            break;
        case TYPE_PAIR:
            printf("(");
            while (obj->type == TYPE_PAIR) {
                print_object(obj->car);
                obj = obj->cdr;
                if (obj->type == TYPE_PAIR) printf(" ");
            }
            if (obj != NIL) {
                printf(" . ");
                print_object(obj);
            }
            printf(")");
            break;
        case TYPE_NIL:
            printf("()");
            break;
        default:
            printf("<unknown>");
            break;
    }
}

bool object_equal(Object *a, Object *b) {
    if (a == b) return true;

    if (a->type != b->type) return false;

    switch (a->type) {
        case TYPE_INT:
            return a->int_val == b->int_val;

        case TYPE_SYMBOL:
            return strcmp(a->symbol, b->symbol) == 0;

        case TYPE_NIL:
            return true;

        case TYPE_PAIR:
            return object_equal(a->car, b->car) &&
                   object_equal(a->cdr, b->cdr);

        default:
            return false;
    }
}

Object *lisp_eq(Object *a, Object *b) {
    return (a == b) ? TRUE : NIL;
}

Object *lisp_equal(Object *a, Object *b) {
    return object_equal(a, b) ? TRUE : NIL;
}

#endif // OBJECT_H
