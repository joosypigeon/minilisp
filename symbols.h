#ifndef SYMBOLS_H
#define SYMBOLS_H

#define SYM_INT     "int"
#define SYM_SYMBOL  "symbol"
#define SYM_LAMBDA  "lambda"
#define SYM_PAIR    "pair"
#define SYM_NIL     "NIL"
#define SYM_TRUE    "#t"
#define SYM_FALSE   "#f"
#define SYM_UNKOWN  "unkown"

#define SYM_DEFINE  "define"
#define SYM_QUOTE   "quote"
#define SYM_EQUAL   "equal?"
#define SYM_ATOM    "atom?"
#define SYM_SET     "set!"
#define SYM_LET     "let"
#define SYM_PRINT   "print"
#define SYM_CONS    "cons"
#define SYM_CAR     "car"
#define SYM_CDR     "cdr"
#define SYM_COND    "cond"
#define SYM_LIST    "list"
#define SYM_BEGIN   "begin"
#define SYM_NUMBER_EQUAL  "="
#define SYM_PLUS    "+"
#define SYM_MINUS   "-"
#define SYM_TIMES   "*"
#define SYM_CHAR_QUOTE  '\''

#define SYM_IF      "if"

#define SYM_SPACE  ' '

#endif // SYMBOLS_H