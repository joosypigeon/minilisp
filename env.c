#include "env.h"
#include "debug.h"

Env *global = NULL;

void env_define(Env *local, const char *name, Object *value) {
    Binding *b = malloc(sizeof(Binding));
    b->name = xstrdup(name);
    b->value = value;
    b->next = local->bindings;
    local->bindings = b;
    DEBUG_PRINT_VERBOSE("env_define: env: %p, %s = %s\n", local, name, object_to_string(value));
}

Object *env_lookup(Env *local, const char *name) {
    for (Env* level = local; level; level = level->parent) {
        for (Binding *b = level->bindings; b != NULL; b = b->next) {
            if (strcmp(b->name, name) == 0) {
                DEBUG_PRINT_VERBOSE("env_lookup: env: %p, %s = %s\n", level, name, object_to_string(b->value));
                return b->value;
            }
        }
    }
    fprintf(stderr, "Unbound symbol: %s\n", name);
    exit(1);
}

Env* push_env(Env *local){
    if(local == NULL){
        DEBUG_PRINT_ERROR("expected enviroment");
        exit(1);
    }
    Env* new = malloc(sizeof(Env));
    new->bindings = NULL;
    new->parent = local;
    DEBUG_PRINT_VERBOSE("push_env: new env: %p, parent: %p\n", new, local);
    return new;
}

Env* pop_env(Env *local){
    Env* parent = local->parent;
    Binding *b = local->bindings;
    while(b){
        Binding* next = b->next;
        free(b->name);
        free(b->value);
        free(b);
        b = next;
    }
    free(local);
    return parent;
}