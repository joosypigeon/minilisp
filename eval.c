#include "eval.h"
#include "debug.h"
#include "symbols.h"



typedef Object *(*SpecialFormHandler)(Env *, Object *);

Object *handle_quote(Env *env, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_quote: env: %p, expr: %s\n", env, object_to_string(expr));
    int length = list_length(expr);
    if (length != 2) {
        DEBUG_PRINT_ERROR("Invalid quote form\n");
        exit(1);
    }
    return cadr(expr);
}

Object *handle_define(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_define: local: %p, expr: %s\n", local, object_to_string(expr));
    int length = list_length(expr);
    if (length != 3) {
        DEBUG_PRINT_ERROR("Invalid define form\n");
        exit(1);
    }
    Object *second = cadr(expr);
    if (second->type == TYPE_SYMBOL) {
        Object *val = eval(local, caddr(expr));
        env_define(local, second->symbol, val);
        return val;
    }
    if(second->type==TYPE_PAIR){
        DEBUG_PRINT_VERBOSE("found function\n");
        Object *name = car(second);
        if(name->type!=TYPE_SYMBOL){
            DEBUG_PRINT_ERROR("handle_define: expected function name\n");
            exit(1);
        }
        Object* parameters = cdr(second);
        Object *body = caddr(expr);
        if(body->type!=TYPE_PAIR){
            DEBUG_PRINT_ERROR("handle_define: expected function body\n");
            exit(1);
        }
        Env *env = push_env(local);
        Object* lambda = make_lambda(parameters, body, env);

        env_define(local, name->symbol, lambda);

        return lambda;
    }

    DEBUG_PRINT_ERROR("Invalid define form\n");
    exit(1);
}

Object *handle_set(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_set: local: %p, expr: %s\n", local, object_to_string(expr));
    int length = list_length(expr);
    if (length != 3) {
        DEBUG_PRINT_ERROR("Invalid define form\n");
        exit(1);
    }
    Object *second = cadr(expr);
    if (second->type == TYPE_SYMBOL) {
        Object *val = eval(local, caddr(expr));
        bool found = set(local, second->symbol, val);
        if (!found) {
            DEBUG_PRINT_ERROR("handle_set: %s not found\n", second->symbol);
            exit(1);
        }
        return val;
    }
    if(second->type==TYPE_PAIR){
        DEBUG_PRINT_VERBOSE("found function\n");
        Object *name = car(second);
        if(name->type!=TYPE_SYMBOL){
            DEBUG_PRINT_ERROR("handle_set: expected function name\n");
            exit(1);
        }
        Object* parameters = cdr(second);
        Object *body = caddr(expr);
        if(body->type!=TYPE_PAIR){
            DEBUG_PRINT_ERROR("handle_set: expected function body\n");
            exit(1);
        }
        Env *env = push_env(local);
        Object* lambda = make_lambda(parameters, body, env);

        bool found = set(local, second->symbol, lambda);
        if (!found) {
            DEBUG_PRINT_ERROR("handle_set: %s not found\n", second->symbol);
            exit(1);
        }

        return lambda;
    }

    DEBUG_PRINT_ERROR("Invalid define form\n");
    exit(1);
}

Object *handle_plus(Env *env, Object *expr) {
    Object *iter = cdr(expr);
    Object *result = make_number(0); // start with 0

    while (iter != NIL) {
        Object *val = eval(env, car(iter));
        if (val->type != TYPE_INT) {
            DEBUG_PRINT_ERROR("+: expected int\n");
            exit(1);
        }
        mpz_add(result->int_val, result->int_val, val->int_val);
        iter = cdr(iter);
    }

    return result;
}

Object *handle_minus(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_minus: local: %p, expr: %s\n", local, object_to_string(expr));
    
    int length = list_length(expr);
    if (length < 2) {
        DEBUG_PRINT_ERROR("Invalid minus form %s\n", object_to_string(expr));
        exit(1);
    }

    Object *first = eval(local, cadr(expr));
    if (first->type != TYPE_INT) {
        DEBUG_PRINT_ERROR("Type error: expected int, found %s\n", object_to_string(first));
        exit(1);
    }

    Object *result = make_number(0);  // initialize result

    if (length == 2) {
        // Unary negation: (- x)
        mpz_neg(result->int_val, first->int_val);
        return result;
    }

    // Start with the first operand
    mpz_set(result->int_val, first->int_val);

    // Iterate through the rest and subtract each
    Object *iter = cddr(expr);
    while (iter != NIL) {
        Object *item = eval(local, car(iter));
        if (item->type != TYPE_INT) {
            DEBUG_PRINT_ERROR("Type error: expected int, got %s\n", object_to_string(item));
            exit(1);
        }

        mpz_sub(result->int_val, result->int_val, item->int_val);
        iter = cdr(iter);
    }

    return result;
}

Object *handle_times(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_times: local: %p, expr: %s\n", local, object_to_string(expr));

    Object *result = make_number(1);  // initialize to 1

    Object *iter = cdr(expr);
    while (iter != NIL) {
        Object *item = eval(local, car(iter));   // evaluate
        if (item->type != TYPE_INT) {
            DEBUG_PRINT_ERROR("Type error: expected int, got %s\n", object_to_string(item));
            exit(1);
        }

        mpz_mul(result->int_val, result->int_val, item->int_val);  // result *= item
        iter = cdr(iter);
    }

    return result;
}

Object *handle_lambda(Env *local, Object *expr) {
    Env *env = push_env(local);
    Object* params = cadr(expr);
    if(params->type!=TYPE_PAIR && params->type!=TYPE_NIL){
        DEBUG_PRINT_ERROR("expected parameters\n");
        exit(1);
    }
    Object *body = caddr(expr);
    Object* lambda = make_lambda(params, body, env);
    return lambda;
}

Object *handle_equal(Env *local, Object *expr){
    DEBUG_PRINT_VERBOSE("enter: handle_equal: local: %p, expr: %s\n", local, object_to_string(expr));
    Object *second = eval(local, cadr(expr));
    Object *third = eval(local, caddr(expr));
    return lisp_equal(second, third);
}

Object *handle_if(Env *env, Object *expr) {
    int length = list_length(expr);
    if (length != 4) {
        DEBUG_PRINT_ERROR("if: wrong number of arguments, expected 3\n");
        exit(1);
    }

    Object *condition = cadr(expr);
    Object *then_branch = caddr(expr);
    Object *else_branch = cadddr(expr);

    Object *cond_result = eval(env, condition);
    if (cond_result != NIL) {
        return eval(env, then_branch);
    } else {
        return eval(env, else_branch);
    }
}

Object *handle_cons(Env *env, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_cons: env: %p, expr: %s\n", env, object_to_string(expr));
    int length = list_length(expr);
    if (length != 3) {
        DEBUG_PRINT_ERROR("Invalid cons form\n");
        exit(1);
    }
    Object *first = eval(env, cadr(expr));
    Object *second = eval(env, caddr(expr));
    Object *result = cons(first, second);
    return result;
}

Object *handle_car(Env *env, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_car: env: %p, expr: %s\n", env, object_to_string(expr));
    int length = list_length(expr);
    if (length != 2) {
        DEBUG_PRINT_ERROR("Invalid car form\n");
        exit(1);
    }
    Object *first = eval(env, cadr(expr));

    if (first->type != TYPE_PAIR) {
        DEBUG_PRINT_ERROR("car: expected pair\n");
        exit(1);
    }
    Object *result = first->car;
    return result;
}

Object *handle_cdr(Env *env, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: handle_cdr: env: %p, expr: %s\n", env, object_to_string(expr));
    int length = list_length(expr);
    if (length != 2) {
        DEBUG_PRINT_ERROR("Invalid cdr form\n");
        exit(1);
    }
    Object *first = eval(env, cadr(expr));
    if (first->type != TYPE_PAIR) {
        DEBUG_PRINT_ERROR("cdr: expected pair\n");
        exit(1);
    }
    Object *result = first->cdr;
    return result;
}

typedef struct {
    const char *symbol;
    SpecialFormHandler handler;
} DispatchEntry;

DispatchEntry special_forms[] = {
    { SYM_QUOTE, handle_quote },
    { SYM_DEFINE, handle_define },
    { SYM_PLUS, handle_plus},
    { SYM_MINUS, handle_minus},
    { SYM_TIMES, handle_times},
    { SYM_LAMBDA, handle_lambda},
    { SYM_EQUAL, handle_equal},
    { SYM_IF, handle_if},
    { SYM_SET, handle_set},
    { SYM_CONS, handle_cons},
    { SYM_CAR, handle_car},
    { SYM_CDR, handle_cdr},
    // ... add more here
    { NULL, NULL }
};




Object *eval(Env *local, Object *expr);

Object *eval_function(Env *env, Object* lambda, Object* args){
    DEBUG_PRINT_VERBOSE("enter: eval_function: lambda: %s, args: %s\n", object_to_string(lambda), object_to_string(args));

    Object* params = lambda->lambda.params;
    //Env *local = push_env(lambda->lambda.env);

    Env *local = push_env(env);

    int params_length = list_length(params);
    int args_length = list_length(args);
    if (params_length != args_length) {
        DEBUG_PRINT_ERROR("TYPE ERROR: expected %i arguments, got %i\n", params_length, args_length);
        exit(1);
    }
    
    while (params != NIL && args != NIL) {
        Object *param = car(params);  // symbol
        Object *arg = eval(local, car(args));      // value
    
        if (param->type != TYPE_SYMBOL) {
            DEBUG_PRINT_ERROR("TYPE ERROR: expected a symbol\n"); 
            exit(1);
        }
        /*
        if (arg->type != TYPE_INT) {
            DEBUG_PRINT_ERROR("TYPE ERROR: expected an integer\n"); 
            exit(1);
        }
*/
        DEBUG_PRINT_VERBOSE("eval_function: param: %s, arg: %s\n", param->symbol, object_to_string(arg));
    
        env_define(local, param->symbol, arg);
    
        params = cdr(params);
        args = cdr(args);
    }

    Object* body = lambda->lambda.body;

    Object* value = eval(local, body);
    DEBUG_PRINT_VERBOSE("leave: eval_function: value: %s\n", object_to_string(value));
    return value;
}

Object *eval_list(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: eval_list: local: %p, expr: %s\n", local, object_to_string(expr));
    Object *first = car(expr);
    if (first->type == TYPE_SYMBOL) {
        for (int i = 0; special_forms[i].symbol != NULL; i++) {
            if (strcmp(first->symbol, special_forms[i].symbol) == 0) {
                return special_forms[i].handler(local, expr);
            }
        }
        /* expect function call*/
        Object *lambda = env_lookup(local, first->symbol);
        if(lambda->type!=TYPE_LAMBDA){
            DEBUG_PRINT_ERROR("TYPE ERROR: expected lambda expression\n");
            exit(1);
        }
        Object* arg_list = cdr(expr);
        Object* value = eval_function(local, lambda, arg_list);
        return value;
    }
    /* Should have a lambda expression*/
    if (first->type == TYPE_PAIR) {
        Object* lambda = eval(local, first);
        if(lambda->type!=TYPE_LAMBDA) {
            DEBUG_PRINT_ERROR("TYPE ERROR: expected lambda expresions, got %s\n", type_to_string(lambda->type));
            exit(1);               
        }
        Object *args = cdr(expr);
        Object* value = eval_function(local, lambda, args);
        return value;
    }

    DEBUG_PRINT_ERROR("Cannot evaluate expression yet\n");
    print_object(expr);
    printf("\n");
    exit(1);
}

Object *eval(Env *local, Object *expr) {
    DEBUG_PRINT_VERBOSE("enter: eval: local: %p, expr: %s\n", local, object_to_string(expr));
    Object *result = expr;
    if (expr->type == TYPE_SYMBOL)
        result = env_lookup(local, expr->symbol);
    if (expr->type == TYPE_PAIR)
        result = eval_list(local, expr);
    DEBUG_PRINT_VERBOSE("leave: eval: expr: %s\n", object_to_string(result));
    return result;
}