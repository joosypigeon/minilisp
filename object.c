#include <stdarg.h>

#include "object.h"
#include "debug.h"
#include "symbols.h"
#include "error.h"

// Singleton NIL object
Object *NIL = &(Object){ .type = TYPE_NIL, .symbol = "#f" };

// Singleton true object
Object *TRUE = &(Object){ .type = TYPE_SYMBOL, .symbol = "#t" };

bool is_pair(Object *item){
    return item->type==TYPE_PAIR;
}

bool is_nil(Object *item){
    return item==NIL;
}

Object *make_number(int n) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_INT;
    mpz_init_set_si(obj->int_val, n); // initialize and set value
    return obj;
}

Object *make_number_from_string(const char *tok) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_INT;
    mpz_init_set_str(obj->int_val, tok, 10);
    if (mpz_cmp_ui(obj->int_val, 0) < 0) {
        RAISE_ERROR("Invalid number: %s\n", tok);
        exit(1);
    }
    return obj;
}

Object *make_symbol(const char *name) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_SYMBOL;
    obj->symbol = xstrdup(name);
    return obj;
}

Object *make_string(const char *value) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_STRING;
    obj->str_val = xstrdup(value);  // use your safe strdup
    return obj;
}

Object *make_lambda(Object *params, Object *body, Env *env) {
    DEBUG_PRINT_VERBOSE("make_lambda: env: %p, params: %s, body: %s\n",
        env, object_to_string(params), object_to_string(body));
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_LAMBDA;
    obj->lambda.params = params;
    obj->lambda.body = body;
    obj->lambda.env = env;
    return obj;
}

Object *make_nil(void) {
    return NIL;
}

Object *make_true(void) {
    return TRUE;
}

Object *make_error(const char *fmt, ...) {
    va_list args;
    va_start(args, fmt);

    char *msg = malloc(256);
    vsnprintf(msg, 256, fmt, args);  // format like printf

    va_end(args);

    Object *err = malloc(sizeof(Object));
    err->type = TYPE_ERROR;
    err->error_msg = msg;

    return err;
}

Object *cons(Object *car, Object *cdr) {
    Object *obj = malloc(sizeof(Object));
    obj->type = TYPE_PAIR;
    obj->car = car;
    obj->cdr = cdr;
    return obj;
}

Object *car(Object *obj) {
    if (obj->type != TYPE_PAIR) {
        RAISE_ERROR("Error: car called on non-pair %s\n", object_to_string(obj));
        exit(1);
    }
    return obj->car;
}

Object *cdr(Object *obj) {
    if (obj->type != TYPE_PAIR) {
        RAISE_ERROR("Error: cdr called on non-pair\n");
        exit(1);
    }
    return obj->cdr;
}

Object *cadr(Object *obj) {
    return car(cdr(obj));
}

Object *cddr(Object *obj) {
    return cdr(cdr(obj));
}

Object *caddr(Object *obj) {
    return car(cdr(cdr(obj)));
}

Object *cadddr(Object *obj) {
    return car(cdr(cdr(cdr(obj))));
}

int list_length(Object *obj) {
    if (obj == NIL) return 0;

    int len = 0;

    while (obj != NIL) {
        if (!obj) {
            RAISE_ERROR("unexpected NULL object in list\n");
            exit(1);
        }

        if (obj->type != TYPE_PAIR) {
            RAISE_ERROR("list_length: not TYPE_PAIR, found %s\n",
                              type_to_string(obj->type));
            exit(1);
        }

        len++;
        obj = obj->cdr;
    }

    return len;
}


char * type_to_string(Type type) {
    switch (type) {
        case TYPE_INT:    return xstrdup(SYM_INT);
        case TYPE_SYMBOL: return xstrdup(SYM_SYMBOL);
        case TYPE_LAMBDA: return xstrdup(SYM_LAMBDA);
        case TYPE_PAIR:   return xstrdup(SYM_PAIR);
        case TYPE_NIL:    return xstrdup(SYM_NIL);
        default:          return xstrdup(SYM_UNKOWN);
    }
}

void print_type(Type type) {
    printf("%s", type_to_string(type));
}

void print_object(Object *obj) {
    char *str = object_to_string(obj);
    printf("%s", str);
    free(str);
}

#define INITIAL_BUF_SIZE 128

static void append_str(char **buf, size_t *size, size_t *used, const char *str) {
    size_t len = strlen(str);
    while (*used + len + 1 > *size) {
        *size *= 2;
        *buf = realloc(*buf, *size);
    }
    strcpy(*buf + *used, str);
    *used += len;
}

static void append_char(char **buf, size_t *size, size_t *used, char c) {
    if (*used + 2 > *size) {
        *size *= 2;
        *buf = realloc(*buf, *size);
    }
    (*buf)[(*used)++] = c;
    (*buf)[*used] = '\0';
}

static void object_to_string_internal(Object *obj, char **buf, size_t *size, size_t *used) {
    switch (obj->type) {
        case TYPE_INT: {
            char *str = mpz_get_str(NULL, 10, obj->int_val);  // base 10 string
            append_str(buf, size, used, str);
            free(str);  // important: GMP allocates it, you must free it
            break;
        }

        case TYPE_SYMBOL:
            append_str(buf, size, used, obj->symbol);
            break;

        case TYPE_STRING:
            append_char(buf, size, used, '"');
            append_str(buf, size, used, obj->str_val);
            append_char(buf, size, used, '"');
            break;

        case TYPE_LAMBDA:
            append_str(buf, size, used, "(lambda ");
            object_to_string_internal(obj->lambda.params, buf, size, used);
            append_char(buf, size, used, ' ');
            object_to_string_internal(obj->lambda.body, buf, size, used);
            append_char(buf, size, used, ')');
            break;

        case TYPE_PAIR:
            append_char(buf, size, used, '(');
            while (obj->type == TYPE_PAIR) {
                object_to_string_internal(obj->car, buf, size, used);
                obj = obj->cdr;
                if (obj->type == TYPE_PAIR)
                    append_char(buf, size, used, ' ');
            }
            if (obj != NIL) {
                append_str(buf, size, used, " . ");
                object_to_string_internal(obj, buf, size, used);
            }
            append_char(buf, size, used, ')');
            break;

        case TYPE_NIL:
            append_str(buf, size, used, "NIL");
            break;

        default:
            append_str(buf, size, used, "<unknown>");
            break;
    }
}

char *object_to_string(Object *obj) {
    size_t size = INITIAL_BUF_SIZE;
    size_t used = 0;
    char *buf = malloc(size);
    buf[0] = '\0';

    object_to_string_internal(obj, &buf, &size, &used);

    return buf;
}

bool object_equal(Object *a, Object *b) {
    DEBUG_PRINT_VERBOSE("object_equal: a: %s, b: %s\n",
        object_to_string(a), object_to_string(b));
    if (a == b) return true;

    if (a->type != b->type) return false;

    switch (a->type) {
        case TYPE_INT:
        if (mpz_cmp(a->int_val, b->int_val) == 0) {
            DEBUG_PRINT_VERBOSE("object_equal: int equal\n");
        }
        return mpz_cmp(a->int_val, b->int_val) == 0;

        case TYPE_SYMBOL:
            return strcmp(a->symbol, b->symbol) == 0;

        case TYPE_STRING:
            return strcmp(a->str_val, b->str_val) == 0;

        case TYPE_NIL:
            return true;

        case TYPE_PAIR:
            return object_equal(a->car, b->car) &&
                   object_equal(a->cdr, b->cdr);

        default:
            return false;
    }
}

Object *lisp_eq(Object *a, Object *b) {
    DEBUG_PRINT_VERBOSE("lisp_eq: a: %s, b: %s\n",
        object_to_string(a), object_to_string(b));
    return (a == b) ? TRUE : NIL;
}

Object *lisp_equal(Object *a, Object *b) {
    DEBUG_PRINT_VERBOSE("lisp_equal: a: %s, b: %s\n",
        object_to_string(a), object_to_string(b));
    return object_equal(a, b) ? TRUE : NIL;
}