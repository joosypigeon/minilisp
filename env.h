#ifndef ENV_H
#define ENV_H

#include "object.h"

typedef struct Binding {
    Object *name;
    Object *value;
    struct Binding *next;
} Binding;

typedef struct Env {
    bool marked;
    Env *next_allocated;
    Binding *bindings;
    struct Env *parent;
} Env;

extern Env *global;

void env_define(Env *local, Object *name, Object *value);
bool set(Env *local, Object *name, Object *value);
Object *env_lookup(Env *local, Object *name);
Env* push_env(Env *local);
Env* pop_env(Env *local);
void free_all_envs();

#endif // ENV_H