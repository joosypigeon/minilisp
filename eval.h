#ifndef EVAL_H
#define EVAL_H  

#include "object.h"
#include "env.h"


Object *safe_eval(Env *local, Object *expr);

#endif // EVAL_H