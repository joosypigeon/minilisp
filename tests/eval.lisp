;; This test file deliberately triggers evaluator errors,
;; and expects them to return error objects (not crash).

;; Attempt to call a non-function
(print (print 42))               ; OK
(print (42 1 2))                 ; ERROR: 42 is not callable

;; Wrong number of args to lambda
(define id (lambda (x) x))
(print (id))                     ; ERROR: missing arg
(print (id 1 2))                 ; ERROR: too many args

;; Type mismatch
(print (+ 1 'x))                 ; ERROR: 'x is not a number
(print (* "hello" 5))           ; ERROR: "hello" is not a number

;; Accessing car of an atom
(print (car 42))                ; ERROR: not a pair
(print (cdr "not a list"))      ; ERROR: not a pair

;; Bad cond form
(print (cond (42)))             ; ERROR: malformed cond clause
(print (cond ((equal? 1 1))))   ; ERROR: missing result in clause

;; Bad define
(print (define 42 99))          ; ERROR: define must use symbol
(print (define (42 x) x))       ; ERROR: function name must be a symbol

;; Unbound symbol
(print unknown-variable)        ; ERROR: unbound symbol
