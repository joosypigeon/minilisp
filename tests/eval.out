42
42
eval.c:420: Cannot evaluate expression.

eval.c:362: TYPE ERROR: expected 1 arguments, got 0

eval.c:362: TYPE ERROR: expected 1 arguments, got 2

eval.c:91: +: expected int

eval.c:146: Type error: expected int, got "hello"

eval.c:207: car: expected pair

eval.c:219: cdr: expected pair

42
#t
eval.c:46: Invalid define form

eval.c:33: handle_define: expected function name

env.c:88: env_lookup: unknown-variable not found

